File: k9mail/src/main/java/com/fsck/k9/account/OAuth2WebViewClient.java
package com.fsck.k9.account;

import android.annotation.TargetApi;
import android.net.Uri;
import android.os.Build;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.fsck.k9.activity.setup.AccountSetupPresenter;

import timber.log.Timber;

/**
 * bass class for standard authorization code flow like Google's or Microsoft's
 */
abstract class OAuth2WebViewClient extends WebViewClient {
    private Oauth2PromptRequestHandler requestHandler;
    private OAuth2ErrorHandler errorHandler;

    public OAuth2WebViewClient(Oauth2PromptRequestHandler requestHandler) {
        this.requestHandler = requestHandler;
    }

    protected abstract boolean arrivedAtRedirectUri(Uri uri);
    protected abstract boolean getOutOfDomain(Uri uri);

    @SuppressWarnings("deprecation")
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        Uri uri = Uri.parse(url);

        if (arrivedAtRedirectUri(uri)) {
            final String error = uri.getQueryParameter("error");
            if (error != null) {
                Timber.i("got oauth error: " + error);
                errorHandler.onError(error);
                requestHandler.onErrorWhenGettingOAuthCode(error);
                return true;
            }

            String oAuthCode = uri.getQueryParameter("code");
            requestHandler.onOAuthCodeGot(oAuthCode);
            return true;
        }

        // if (!uri.getHost().contains("google")) {
        if (getOutOfDomain(uri)) {
            requestHandler.onErrorWhenGettingOAuthCode("Don't surf away"); // TODO: 2017/8/19 better error message
            return true;
        }
        return false;
    }

    @Override
    @TargetApi(Build.VERSION_CODES.N)
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        return shouldOverrideUrlLoading(view, request.getUrl().toString());
    }
}

File: mail/common/src/main/java/com/fsck/k9/mail/oauth/authorizationserver/codegrantflow/OAuth2WebViewClient.kt
package com.fsck.k9.mail.oauth.authorizationserver.codegrantflow

import android.annotation.TargetApi
import android.net.Uri
import android.os.Build
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.fsck.k9.mail.AuthenticationFailedException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber

/**
 * Bass class for standard authorization code flow like Google's or Microsoft's
 */
abstract class OAuth2WebViewClient(
    private val email: String,
    private val codeGrantFlowManager: OAuth2CodeGrantFlowManager
) : WebViewClient() {
    protected abstract fun arrivedAtRedirectUri(uri: Uri?): Boolean

    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        return handleUrl(url)
    }

    @TargetApi(Build.VERSION_CODES.N)
    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        return handleUrl(request.url.toString())
    }

    private fun handleUrl(url: String): Boolean {
        Timber.d(url)
        val uri = Uri.parse(url)
        if (arrivedAtRedirectUri(uri)) {
            val error = uri.getQueryParameter("error")
            if (error != null) {
                Timber.e("got oauth error: $error")
                codeGrantFlowManager.promptRequestHandler?.onError(error)
                return true
            }

            uri.getQueryParameter("code")?.let { oauthCode ->
                codeGrantFlowManager.promptRequestHandler?.onObtainCodeSuccessful()
                GlobalScope.launch {
                    withContext(Dispatchers.IO) {
                        try {
                            codeGrantFlowManager.exchangeCode(email, oauthCode)
                            withContext(Dispatchers.Main) {
                                codeGrantFlowManager.promptRequestHandler?.onObtainAccessTokenSuccessful()
                            }
                        } catch (e: AuthenticationFailedException) {
                            withContext(Dispatchers.Main) {
                                codeGrantFlowManager.promptRequestHandler?.onError("Error when exchanging code")
                            }
                        }
                    }
                }
            } ?: run {
                codeGrantFlowManager.promptRequestHandler?.onError("No auth code")
            }
            return true
        }
        return false
    }
}
